File: ai/neural_network.py
import numpy as np

class NeuralNetwork:
    def __init__(self, input_size=3, hidden_size=6, output_size=1):
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.output_size = output_size
        
        # Initialize weights with smaller values
        self.weights1 = np.random.randn(input_size, hidden_size) * 0.1
        self.weights2 = np.random.randn(hidden_size, output_size) * 0.1
        self.bias1 = np.random.randn(hidden_size) * 0.1
        self.bias2 = np.random.randn(output_size) * 0.1
    
    def sigmoid(self, x):
        return 1 / (1 + np.exp(-x))
    
    def forward(self, inputs):
        # Convert inputs to numpy array
        x = np.array(inputs)
        
        # Hidden layer
        hidden = self.sigmoid(np.dot(x, self.weights1) + self.bias1)
        
        # Output layer
        output = self.sigmoid(np.dot(hidden, self.weights2) + self.bias2)
        
        return output[0]  # Return scalar value
    
    def get_weights(self):
        return {
            'w1': self.weights1,
            'w2': self.weights2,
            'b1': self.bias1,
            'b2': self.bias2
        }
    
    def set_weights(self, weights):
        self.weights1 = weights['w1']
        self.weights2 = weights['w2']
        self.bias1 = weights['b1']
        self.bias2 = weights['b2'] 
    
    def copy(self):
        """Create a deep copy of the neural network"""
        new_network = NeuralNetwork(self.input_size, self.hidden_size, self.output_size)
        weights = self.get_weights()
        new_network.set_weights({
            'w1': weights['w1'].copy(),
            'w2': weights['w2'].copy(),
            'b1': weights['b1'].copy(),
            'b2': weights['b2'].copy()
        })
        return new_network

File: static/js/neural_network.js
class NeuralNetwork {
    constructor(inputSize = 3, hiddenSize = 8, outputSize = 1) {
        this.inputSize = inputSize;
        this.hiddenSize = hiddenSize;
        this.outputSize = outputSize;
        
        // Initialize weights and biases
        this.weights1 = this.createRandomMatrix(inputSize, hiddenSize);
        this.weights2 = this.createRandomMatrix(hiddenSize, outputSize);
        this.bias1 = this.createRandomArray(hiddenSize);
        this.bias2 = this.createRandomArray(outputSize);
        
        // Verify initialization
        if (!this.weights1 || !this.weights2) {
            throw new Error('Failed to initialize neural network weights');
        }
    }

    createRandomMatrix(rows, cols) {
        return Array.from({ length: rows }, () =>
            Array.from({ length: cols }, () => (Math.random() * 2 - 1) * 0.5)
        );
    }

    createRandomArray(size) {
        return Array.from({ length: size }, () => (Math.random() * 2 - 1) * 0.5);
    }

    sigmoid(x) {
        return 1 / (1 + Math.exp(-Math.max(-10, Math.min(10, x))));
    }

    forward(inputs) {
        const normalizedInputs = [...inputs];
        
        // Clamp inputs to [-1, 1]
        normalizedInputs.forEach((val, i) => {
            normalizedInputs[i] = Math.max(-1, Math.min(1, val));
        });
        
        // Hidden layer
        const hidden = Array(this.hiddenSize).fill(0);
        for (let i = 0; i < this.hiddenSize; i++) {
            let sum = this.bias1[i];
            for (let j = 0; j < this.inputSize; j++) {
                sum += normalizedInputs[j] * this.weights1[j][i];
            }
            hidden[i] = this.sigmoid(sum);
        }

        // Output layer
        let output = this.bias2[0];
        for (let i = 0; i < this.hiddenSize; i++) {
            output += hidden[i] * this.weights2[i][0];
        }
        
        // Single sigmoid activation
        const finalOutput = this.sigmoid(output);
        
        // Store for visualization
        this.lastHiddenActivations = hidden;
        this.lastOutput = finalOutput;
        
        return finalOutput;
    }

    copy() {
        const newNetwork = new NeuralNetwork(this.inputSize, this.hiddenSize, this.outputSize);
        newNetwork.weights1 = this.weights1.map(row => [...row]);
        newNetwork.weights2 = this.weights2.map(row => [...row]);
        newNetwork.bias1 = [...this.bias1];
        newNetwork.bias2 = [...this.bias2];
        return newNetwork;
    }
} 